data_model/src/predicate.rs

```diff
@@ -0,0 +1,852 @@
+        PredicateBox::Raw(value::Predicate::Pass)
+    }
+}
+
+impl PredicateTrait<Value> for PredicateBox {
+    #[inline] // This is not a simple function, but it allows you to inline the logic and optimise away the logical operations.
+    fn applies(&self, input: &Value) -> bool {
+        match self {
+            PredicateBox::And(vector) => vector.iter().all(|pred| pred.applies(input)),
+            PredicateBox::Or(vector) => vector.iter().any(|pred| pred.applies(input)),
+            PredicateBox::Not(predicate) => !predicate.applies(input),
+            PredicateBox::Raw(predicate) => predicate.applies(input),
+        }
+    }
+}
+
+#[cfg(test)]
+#[allow(clippy::print_stdout, clippy::use_debug)]
+pub mod test {
+    use super::{value, PredicateBox};
+    use crate::{PredicateTrait as _, Value};
+
+    #[test]
+    fn pass() {
+        let t = PredicateBox::new(value::Predicate::Pass);
+        let f = t.clone().negate();
+        let v_t = Value::from(true);
+        let v_f = Value::from(false);
+        println!("t: {t:?}, f: {f:?}");
+
+        assert!(t.applies(&v_t));
+        assert!(t.applies(&v_f));
+        assert!(!f.applies(&v_t));
+        assert!(!f.applies(&v_f));
+    }
+
+    #[test]
+    fn truth_table() {
+        let t = PredicateBox::new(value::Predicate::Pass);
+        let f = t.clone().negate();
+        let v = Value::from(true);
+
+        assert!(!PredicateBox::and(t.clone(), f.clone()).applies(&v));
+        assert!(PredicateBox::and(t.clone(), t.clone()).applies(&v));
+        assert!(!PredicateBox::and(f.clone(), f.clone()).applies(&v));
+        assert!(!PredicateBox::and(f.clone(), t.clone()).applies(&v));
+
+        assert!(PredicateBox::or(t.clone(), t.clone()).applies(&v));
+        assert!(PredicateBox::or(t.clone(), f.clone()).applies(&v));
+        assert!(PredicateBox::or(f.clone(), t).applies(&v));
+        assert!(!PredicateBox::or(f.clone(), f).applies(&v));
+    }
+
+    #[test]
+    fn negation() {
+        let t = PredicateBox::default();
+
+        assert!(matches!(t.clone().negate().negate(), PredicateBox::Raw(_)));
+        // De-morgan identities
+        assert!(matches!(
+            PredicateBox::and(t.clone(), t.clone()).negate(),
+            PredicateBox::Or(_)
+        ));
+        assert!(matches!(
+            PredicateBox::or(t.clone(), t).negate(),
+            PredicateBox::And(_)
+        ));
+    }
+}
+
+pub mod string {
+    //! String-related predicates and implementations.
+    use super::*;
+
+    /// Predicate useful for processing [`String`]s and [`Name`]s.
+    #[derive(Debug, Clone, Serialize, Deserialize, Encode, Decode, IntoSchema)]
+    pub enum Predicate {
+        /// Forward to [`str::contains()`]
+        Contains(String),
+        /// Forward to [`str::starts_with()`]
+        StartsWith(String),
+        /// Forward to [`str::ends_with()`]
+        EndsWith(String),
+        /// Forward to [`String`] equality.
+        Is(String),
+    }
```

Comment:
alternative design could have been:
```rs
    pub enum PredicateType {
        Contains,
        StartsWith,
        EndsWith,
        Is,
    }
    pub struct Predicate {
        type_: PredicateType,
        operates_on: String,
    }
```

I'm not saying it would be better, just noting. It is good either way

Reply:
It generates the exact same assembly, but now you can access the string without destructuring the `Predicate`. You are also forced to make all variants of the same type. Also, you have one fewer struct in the namespace. 

I'm not saying it would be worse, just nothing that I considered it, and decided to go with the algebraic form.